micro_dl/inference/stitch_predictions.py

```diff
@@ -0,0 +1,261 @@
+        # initialize all indices to :
+        idx_in_img = []
+        idx_in_block = []
+        for dim_idx in range(len(pred_image.shape)):
+            idx_in_img.append(np.s_[:])
+            idx_in_block.append(np.s_[:])
+        idx_in_img[z_dim] = np.s_[start_idx + num_overlap: end_idx]
+        idx_in_block[z_dim] = np.s_[num_overlap:]
+        pred_image[idx_in_img] = pred_block[idx_in_block]
+        if start_idx > 0:
+            for sl_idx in range(num_overlap):
+                idx_in_img[z_dim] = start_idx + sl_idx
+                idx_in_block[z_dim] = sl_idx
+                if overlap_operation == 'mean':
+                    pred_image[idx_in_img] = \
+                        (reverse_wts[sl_idx] * pred_image[idx_in_img] +
+                         forward_wts[sl_idx] * pred_block[idx_in_block])
+                elif overlap_operation == 'any':
+                    pred_block[idx_in_img] = np.any(pred_image[idx_in_img],
+                                                    pred_block[idx_in_block])
+        else:
+            idx_in_img[z_dim] = np.s_[start_idx: start_idx + num_overlap]
+            idx_in_block[z_dim] = np.s_[0: num_overlap]
+            pred_image[idx_in_img] = pred_block[idx_in_block]
+        return pred_image
+
+    def _stitch_along_z(self,
+                        tile_imgs_list,
+                        block_indices_list):
+        """Stitch images along Z with or w/o overlap
+
+        :param list tile_imgs_list: list with predicted tensors
+        :param list block_indices_list: list with tuples of (start, end) idx
+        :return np.array stitched_img: 3D image with blocks assembled in place
+        """
+
+        stitched_img = np.zeros(self.shape_3d)
+
+        if 'overlap_shape' in self.overlap_dict:
+            assert isinstance(self.overlap_dict['overlap_shape'], int), \
+                'tile_z only supports an overlap of int slices along z'
+
+        for idx, sub_block in enumerate(tile_imgs_list):
+            try:
+                cur_sl_idx = block_indices_list[idx]
+                stitched_img = self._place_block_z(pred_block=sub_block,
+                                                   pred_image=stitched_img,
+                                                   start_idx=cur_sl_idx[0],
+                                                   end_idx=cur_sl_idx[1])
+            except Exception as e:
+                raise Exception('error in _stitch_along_z:{}'.format(e))
+        return stitched_img
+
+    def _place_block_xyz(self,
+                         pred_block,
+                         pred_image,
+                         crop_index):
+        """Place the current block prediction in the larger vol
+
+        :param np.array pred_block: current prediction block
+        :param np.array pred_image: full 3D prediction image with zeros
+        :param list crop_index: tuple of len 6 with start, end indices for
+         three dimensions
+        """
```

Comment:
The return parameter is not mentioned, but as in the previous case it might not be necessary?